lxd/storage/drivers/driver_alletra_wsapi1.go

```diff
@@ -11,3 +41,281 @@ func newAlletraClient(driver *alletra) *alletraClient {
+	// Extract scheme and host from the gateway URL.
+	scheme, host, found := strings.Cut(p.driver.config["alletra.wsapi.url"], "://")
+	if !found {
+		return fmt.Errorf("Invalid HPE Storage WSAPI URL: %q", p.driver.config["alletra.wsapi.url"])
+	}
+
+	// Set request URL scheme and host.
+	url.Scheme = scheme
+	url.Host = host
+
+	var err error
+	var reqBodyReader io.Reader
+
+	if reqBody != nil {
+		reqBodyReader, err = createBodyReader(reqBody)
+		if err != nil {
+			return err
+		}
+	}
+
+	req, err := http.NewRequest(method, url.String(), reqBodyReader)
+	if err != nil {
+		return fmt.Errorf("Failed to create request: %w", err)
+	}
+
+	// Set custom request headers.
+	for k, v := range reqHeaders {
+		req.Header.Add(k, v)
+	}
+
+	req.Header.Add("Accept", "application/json")
+	if reqBody != nil {
+		req.Header.Add("Content-Type", "application/json")
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: shared.IsFalse(p.driver.config["alletra.wsapi.verify"]),
+			},
+		},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Failed to send request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	var responseBodyBuffer bytes.Buffer
+	teeReader := io.TeeReader(resp.Body, &responseBodyBuffer)
+	bodyBytes, err := io.ReadAll(teeReader)
+	if err != nil {
+		return fmt.Errorf("Failed to read response body for TeeReader: %w", err)
+	}
+
+	if resp.Header.Get("Content-Type") != "application/json" && len(bodyBytes) > 0 {
+		return fmt.Errorf("Response Content-type: %q. Only application/json is allowed for non-empty response body", resp.Header.Get("Content-Type"))
+	}
+
+	if resp.StatusCode >= 400 {
+		respBody = &hpeError{}
+		err := json.Unmarshal(bodyBytes, respBody)
+		if err != nil {
+			return fmt.Errorf("HPE failed to parse WSAPI error response: %w", err)
+		}
+	}
+
+	// Extract the response body if requested.
+	if len(bodyBytes) > 0 {
+		err = json.Unmarshal(bodyBytes, &respBody)
+		if err != nil {
+			return fmt.Errorf("HPE failed to read response body from %q: %w", url.String(), err)
+		}
+	}
+
+	// Extract the response headers if requested.
+	if respHeaders != nil {
+		for k, v := range resp.Header {
+			respHeaders[k] = strings.Join(v, ",")
+		}
+	}
+
+	// Return the formatted error from the body
+	hpeErr, assert := respBody.(*hpeError)
+	if assert {
+		hpeErr.HTTPStatusCode = resp.StatusCode
+
+		// The unauthorized error is reported when an invalid (or expired) access token is provided.
+		// Wrap unauthorized requests into an API status error to allow easier checking for expired
+		// token in the requestAuthenticated function.
+		if resp.StatusCode == http.StatusForbidden && hpeErr.Code == 6 { // 6 = INV_SESS_KEY
+			return api.StatusErrorf(http.StatusForbidden, "Unauthorized request")
+		}
+
+		return hpeErr
+	}
+
+	return nil
+}
+
+var sessionKeys = make(map[string]string)
```

Comment:
what clears old entries from this map?

also, related, as map memory allocations are never released AFAIK, we should seek to make each map entry as small as possible, so rather than `string` it could be `*string`.